### PR TITLE
test: btt tests for BNum

### PR DESCRIPTION
**Pull request by wei3erHase**
Closes BAL-132 

- mixed feelings about using `BONE` when not needed, since it makes it hard to reach numbers as `1.25e18` (would be `5 * BONE / 4`), or express precise results, such as `1.25 * 1.25 = 1.5625` (would be... ?)
- removing comments such as `// Action` and `// Post Condition` as they appear differently ordered and is hard to standarize across all tests, i think comments should be just the ones generated by the tree
- keeping BONE for bounds and assumptions, as it make them easier to logic around
- created a manual smock of BNum, basically making each method virtual, running smock, rollbacking the virtual, and deleting all superfluous smock methods (that wouldn't compile bc overriding non-virtual method). i think actually changing BNum to virtual just for mocking doesn't make sense, as smock should support externalizing internal methods

### CODE REVIEW

File: test/unit/BNum.tree [hunk 1]
```diff
@@ -93,5 +93,5 @@ BNumTest::bpow
 │   └── it should revert
 ├── when passing base gte MAX_BPOW_BASE
 │   └── it should revert
-└── when passing known values // generic for some given values
+└── when passing known values
```
Comment (0xteddybear) [inline test/unit/BNum.tree:96]: opinions on hard-coding the values themselves here as well?

File: test/unit/BNum.t.sol [hunk 2]
```diff
@@ -3,379 +3,342 @@ pragma solidity 0.8.25;
...
     assertEq(_result, BONE);
   }
 
   function test_BpowiWhenPassingKnownValues() external {
     // it should return correct value
-    vm.skip(true);
+    //     4. ^ 12 = 16777216
+    uint256 _a = 4e18;
+    uint256 _b = 12;
+
```
Comment (0xteddybear) [inline test/unit/BNum.t.sol:304]: nit: for small & simple cases I'd skip the whitespace separating actions and {pre,post}conditions

**Review by 0xteddybear — CHANGES_REQUESTED** [review]
I think this PR could be merged against `dev` instead, similar to #120 , which would be a simpler approach since we would not be incurring the tech debt of having failing & outdated tests (for BPool, mostly) on a branch we're merging to, which are not being caught by CI because they don't match `Unit` or `Integration` :fearful: 

see #124

other than that, tests look clean and bulloak doesn't complain

File: test/smock/MockBNum.sol [hunk 3]
```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
```
[the remark this answers is not in the document]
Comment (wei3erHase) [reply test/smock/MockBNum.sol:1]: good catch

File: test/unit/BNum.tree (hunk 1 above)
Comment (wei3erHase) [reply test/unit/BNum.tree:96]: mhmmm too much noise? i think we need to leave the tree with as less implementation details as possible

File: test/unit/BNum.t.sol (hunk 2 above)
Comment (wei3erHase) [reply test/unit/BNum.t.sol:304]: doesn't hurt tho 😇

**Comment by wei3erHase** [comment]
Closing in favour of #126

**Comment by wei3erHase** [comment]
Closing in favour of #126

**Pull request closed without merging**